retchmail.1
.TH RETCHMAIL 1 "July 2007" "Retchmail"

.SH NAME
retchmail \- The world's most stupidly fast POP3 retriever.


.SH SYNOPSIS
.B retchmail
.BR \-d \ |
.BR \-dd \ |
.BR \-q \ |
.BR \-qq \ |
.BR \-c \ |
.BR \-V \ |
.BR \-E \ |
.BR \-F \ |
.br
.ti 17
.BI \-t\  deliverto\fR\ |
.IR fromacct ..."

.SH DESCRIPTION
.B retchmail
is a stupidly fast mail retriever, similar to the
.BR fetchmail (1)
program, except that it is blazingly fast, and only handles POP3 and POP3s
(POP3 over SSL).
.PP
When
.B retchmail
starts, it first loads its configuration from
.hy 0
.IR ~/.retchmail/retchmail.conf ,
.hy 14
which contains basic information about the accounts, mailboxes, etc.
.PP
It will then go out and retrieve all of the mail from those accounts. 
.PP
.SH OPTIONS
Several options are recognized by
.BR retchmail .
.TP
.B \-q
Run in quiet mode (Don't print every message header)
.TP
.B \-qq
Run in REALLY quiet mode (only print error messages)
.TP
.B \-c \fR[\|\fIconfig_moniker\fR\|]
Run retchmail with \fIconfig_moniker\fR as the configuration file (instead of
ini:$HOME/.retchmail/retchmail.conf). 
.TP
.B \-E
Explode a multidrop mailbox and deliver mail to the users that X-Envelope-To
says to.
.TP
.B \-F
Flush all messages after downloading - you probably want to make sure this
is on the command line, or else you will keep downloading the same messages
over and over and over and (you get the idea).
.TP
.B \-V
Prints the current
.B retchmail
version information and exits.
.TP
.B \-d
Displays debug information for troubleshooting
.BR retchmail
.TP
.B \-dd
Displays a lot of debug information for troubleshooting
.BR retchmail.
Use this option when reporting any bugs or problems with the program. 
.PP
.SH BUGS
Probably very few, but if you find any, please report them to us. See
the retchmail web page for more information:
.PP
.BR http://alumnit.ca/wiki/index.php?page=RetchMail 
.PP
.SH FILES
.TP
$HOME/.retchmail/retchmail.conf
Configuration file which contains account, passwords and delivery
information. See retchmail.conf(5).
.PP
.SH AUTHORS
Retchmail was originally written by Avery Pennarun and Patrick Patterson. Many people
have later contributed to it.
.PP
.SH SEE ALSO
.BR retchmail.conf (5),
.PP


debian/changelog
retchmail (1:1.1-1) unstable; urgency=low

  * New upstream version.  (Closes: #407873)

 -- William Lachance <wrlach@gmail.com>  Thu, 19 Jul 2007 19:30:09 -0400

retchmail (1:1.0-1) unstable; urgency=low

  * New upstream version.

 -- Simon Law <sfllaw@debian.org>  Tue, 22 Nov 2005 20:58:10 -0500

retchmail (1:0.09-3) unstable; urgency=low

  * Update config documentation. (Closes: #302884)

 -- James Morrison <phython@debian.org>  Sun,  3 Apr 2005 13:19:30 -0400

retchmail (1:0.09-2) unstable; urgency=low

  * Only use low bits of WvLog pointer in WvHash function. (Closes: #302372)
  * Use xplc 0.3.11.

 -- James Morrison <phython@debian.org>  Sat,  2 Apr 2005 11:23:20 -0500

retchmail (1:0.09-1) unstable; urgency=low

  * New upstream release.  (Closes: #290999)

 -- James Morrison <phython@debian.org>  Wed, 22 Mar 2004 22:16:03 -0500

retchmail (1:0.04-2) unstable; urgency=low

  * Don't call callbacks when they are empty. (Closes: #232469)

 -- James Morrison <phython@debian.org>  Wed,  3 Mar 2004 17:16:03 -0500

retchmail (1:0.04-1) unstable; urgency=low

  * New upstream release.

 -- James A. Morrison <phython@debian.org>  Sun, 15 Feb 2004 13:34:48 -0500

retchmail (1:0.03+20031110-1) unstable; urgency=low

  * Fix Build-Depends.

 -- James Morrison <phython@debian.org>  Mon, 10 Nov 2003 10:17:17 -0500

retchmail (1:0.03+20031017-1) unstable; urgency=low

  * Synchronize with upstream CVS.

 -- James A. Morrison <phython@debian.org>  Fri, 17 Oct 2003 17:48:36 -0400

retchmail (0.3) unstable; urgency=high

  * Synchronize with libwvstreams 3.71

  * Make it so that we handle email addresses with @ in them

  * Handle APOP thanks to a patch from markj@cloaked.freeserve.co.uk

  * Fixes core-dump on certain messages

  * Debian Packaging Fixes (Closes: #142527, #142567)

 -- Patrick Patterson <ppatters@nit.ca>  Mon, 29 Jul 2002 18:39:51 -0400

retchmail (0.2) unstable; urgency=low

  * Synchronise with libwvstreams

  * Debian Packaging Fixes

  * Initial Debian release (Closes: #138948)

 -- Patrick Patterson <ppatters@debian.org>  Mon, 28 Jan 2002 14:01:04 -0500

retchmail (0.1) unstable; urgency=low

  * Initial Release.

 -- Patrick Patterson <ppatters@nit.ca>  Mon, 28 Jan 2002 14:01:04 -0500



retchmail.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2005 Net Integration Technologies, Inc.
 *
 * Avery's insanely fast alternative to Fetchmail
 *
 * This code is LGPL - see the file COPYING.LIB for a full copy of the
 * license.
 */

// WvStreams headers
#include "strutils.h"
#include "uniconfroot.h"
#include "wvlockfile.h"
#include "wvver.h"
#include "wvtcp.h"
#include "wvlog.h"
#include "wvpipe.h"
#include "wvistreamlist.h"
#include "wvlogrcv.h"
#include "wvsslstream.h"
#include "wvhashtable.h"
#include "wvargs.h"
#include "wvcrash.h"
#include "wvhex.h"

// Retchmail headers
#include "wvsendmail.h"
#include "wvpopclient.h"

#include <signal.h>
#include <assert.h>
#include <pwd.h>
#include <sys/types.h>

struct LogNum
{
    int num;
    WvString src;
};

DeclareWvDict(LogNum, WvString, src);


class RetchLog : public WvLogConsole
{
public:
    RetchLog(WvLog::LogLevel _max_level) 
	: WvLogConsole(1, _max_level), lognums(5)
	{ maxnum = 0; }
    virtual ~RetchLog() { }
    
    virtual void _begin_line();
    
    int maxnum;
    LogNumDict lognums;
};


void RetchLog::_begin_line()
{
    if (!strncmp(last_source, "PopRetriever ", 13))
    {
	LogNum *lognum = lognums[last_source];
	if (!lognum)
	{
	    lognum = new LogNum;
	    lognum->num = maxnum++;
	    lognum->src = last_source;
	    lognums.add(lognum, true);
	}
	
	// identify the connection without being too verbose
	if (lognum->num < 26)
	{
	    char str[2];
	    str[0] = 'A' + lognum->num;
	    str[1] = 0;
	    write(str);
	}
	else
	    write(lognum->num);
	if (last_level <= WvLog::Info)
	    write("* ");
	else
	    write("  ");
    }
    else
	WvLogConsole::_begin_line(); // just print the whole thing
}



static volatile bool want_to_die = false;
void signal_handler(int signum)
{
    fprintf(stderr, "\nCaught signal %d; cleaning up and terminating.\n",
	    signum);
    want_to_die = true;
    signal(signum, SIG_DFL);
}


static WvPopClient *newpop(WvStringParm acct,
			   WvStringParm _pass, WvStringParm _deliverto,
                           WvStringParm _mda, bool flush, bool apop_en,
                           bool apop_fall_en, bool explode, bool safemode,
			   bool ignorerp)

{
    WvString user(acct), serv, pass(_pass), deliverto(_deliverto),
            mda(_mda);
    bool ssl = false;
    
    char *cptr = strrchr(user.edit(), '@');
    if (cptr)
    {
	*cptr++ = 0;
	serv = cptr;
    }
    else
	serv = "localhost";
    
    cptr = strchr(serv.edit(), ':');
    if (!cptr)
    {
	serv.append(":110");
	cptr = strchr(serv.edit(), ':'); // guaranteed to work now!
    }
    
    if (atoi(cptr+1) == 995)
	ssl = true;
    
    WvTCPConn *tcp = new WvTCPConn(serv);
    WvStream *conn = tcp;
    if (ssl) // FIXME: ssl verify should probably be set to something.
	conn = new WvSSLStream(tcp, NULL);
    
    return new WvPopClient(conn, acct, pass, deliverto, mda, flush, apop_en, apop_fall_en, explode, safemode, ignorerp);
}


static bool dec_log_level_cb(void *userdata)
{
    WvLog::LogLevel &level = *static_cast<WvLog::LogLevel *>(userdata);
    if ((int)level > (int)WvLog::Critical)
	level = (WvLog::LogLevel)((int)level - 1);
    return true;
}


static bool inc_log_level_cb(void *userdata)
{
    WvLog::LogLevel &level = *static_cast<WvLog::LogLevel *>(userdata);
    if ((int)level < (int)WvLog::Debug5)
	level = (WvLog::LogLevel)((int)level + 1);
    return true;
}


int main(int argc, char **argv)
{
    bool flush = false, explode = false, single_stream = false;
    WvLog::LogLevel lvl = WvLog::Debug1;
    WvString deliverto("");
    WvString confmoniker("");
    
    // Set up WvCrash
    wvcrash_setup(argv[0]);
    
    // make sure electric fence works
    free(malloc(1));

    signal(SIGPIPE, SIG_IGN);
    
    struct passwd *pw = getpwuid(getuid());
    if (pw)
	deliverto = pw->pw_name;

    WvArgs args;

    args.add_option('q', "quiet",
		    "Decrease log level (can be used multiple times)",
		    dec_log_level_cb, &lvl);
    
    args.add_option('d', "debug",
		    "Increase log level (can be used multiple times)",
		    inc_log_level_cb, &lvl);

    args.add_set_bool_option('F', "flush", 
			     "Flush (delete) messages after downloading",
			     flush);

    args.add_set_bool_option('E', "explode", 
			     "Send mail to the user on the system corresponding to the user the mail is sent to",
			     explode);
    
    args.add_option('t', "to", 
		    WvString("Deliver mail to <deliverto> (default '%s')", deliverto),
		    "deliverto", deliverto);

    args.add_option('c', "config",
		    "Use <moniker> instead of ini:~/.retchmail/retchmail",
		    "moniker", confmoniker);
    
    args.set_version("Retchmail version " RETCHMAIL_VER_STRING);
    args.set_email("<wvstreams-dev@lists.nit.ca>");

    args.add_set_bool_option('s', "singlestream", 
			     "Use only one TCP stream",
			     single_stream);
    
    WvStringList arguments;
    if (!args.process(argc, argv, &arguments))
	return 255;
    
    WvString lockname("/tmp/retchmail.%s.pid", getlogin());
    WvLockFile lockfile(lockname);

    if (!lockfile.lock())
    {
        if (lockfile.readpid() == -1)
            fprintf(stderr, "retchmail: can't access lockfile at %s.\n",
		    lockname.cstr());
        else
            fprintf(stderr, "retchmail: already running (%i).\n",
		    lockfile.readpid());
        return 3;
    }

    if (!deliverto)
    {
	fprintf(stderr, "retchmail: can't get username for uid#%u.  "
		"You must give the -t option.\n", getuid());
	return 3;
    }
    
    RetchLog logrcv(lvl);
    
    if (!confmoniker)
    {
	// attempting to use $HOME/.retchmail/retchmail.conf\n");
        confmoniker = "ini:";
        confmoniker.append(getenv("HOME"));
        confmoniker.append("/.retchmail/retchmail.conf");
    }

    UniConfRoot cfg(confmoniker);

    if (!cfg.haschildren())
    {
        wvcon->print("No data found, aborting\n");
	exit(1);
    }

    WvPopClient *cli = NULL;
    bool apop_enable = cfg["retchmail"]["Enable APOP"].getmeint(0);
    bool apop_enable_fallback = cfg["retchmail"]["Enable APOP Fallback"].getmeint(0);
    bool safemode = cfg["retchmail"]["Safe Mode"].getmeint(0);
    bool ignorerp = cfg["retchmail"]["Ignore Return-Path"].getmeint(0);
  
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    if (arguments.isempty())	    
    {
        UniConf sect = cfg["POP Servers"];
	if (sect.haschildren())
	{
	    UniConf::Iter i(sect);
	    for (i.rewind(); i.next(); )
	    {
		cli = newpop(i->key(), i->getme(),
			     cfg["POP Targets"][i->key()].getme(deliverto),
                             cfg["MDA Override"][i->key()].getme(
				     "/usr/sbin/sendmail"),
                             flush, apop_enable, apop_enable_fallback,
			     explode ? : cfg["Explode"][i->key()].getmeint(0),
			     safemode, ignorerp);
		WvIStreamList::globallist.append(cli, true, "client");
                if (single_stream)
                    while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	                WvIStreamList::globallist.runonce();
	    }
	}
	else
	{
	    fprintf(stderr, "\n-- No config file and no accounts given on the "
		    "command line!\n");
	    args.print_usage(argc, argv);
	}
    }
    else
    {
	while (!arguments.isempty());
	{
	    WvString user(arguments.popstr());
	    WvString pass = cfg["POP Servers"][user].getme();
	    if (!pass)
	    {
		wvcon->print("Password for <%s>: ", user);
		
		system("stty -echo 2>/dev/null");
		pass = wvcon->blocking_getline(-1);
		system("stty echo 2>/dev/null");
		
		wvcon->print("\n");
	    }
	    
	    cli = newpop(user, pass,
			 cfg["POP Targets"][user].getme(deliverto),
			 cfg["MDA Override"][user].getme(
				 "/usr/sbin/sendmail"),
			 flush, apop_enable, apop_enable_fallback, explode,
			 safemode, ignorerp);
	    WvIStreamList::globallist.append(cli, true, "client");
            if (single_stream)
                while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	            WvIStreamList::globallist.runonce();
	}
    }
    
    if (!single_stream)
        while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	    WvIStreamList::globallist.runonce();

    lockfile.unlock();
}
